### PR TITLE
An inference engine for breast density classification

### PR DESCRIPTION
**Pull request by vikashg**
…ks but I am not sure where the output of the inference is.

Signed-off-by: gupta <gupta@DIGI1036734.mayo.edu>

Fixes # .

### Description
This is a new pull request for a breast density classification. As of now I have written the inference engine so it can do the inference. However, I am not sure where the output of the inference is stored. I can use some help @Nic-Ma @yiheng-wang-nv 

### Status
**Work in progress**

### Please ensure all the checkboxes:

- [x] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [x] In-line docstrings updated.
- [ ] Update `version` and `changelog` in `metadata.json` if changing an existing bundle.
- [x ] Please ensure the naming rules in config files meet our requirements (please refer to: `CONTRIBUTING.md`).
- [ x] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `metadata.json`.
- [ ] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [x] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.
- [x] Avoid using path that contains personal information within config files (such as use `/home/your_name/` for `"bundle_root"`).

### CODE REVIEW

File: models/breast_density_classification/scripts/script.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+import os
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/scripts/script.py:1]: Hi @vikashg , thanks for making this new PR.
Here are some code format issues (https://github.com/Project-MONAI/model-zoo/actions/runs/3209554960/jobs/5246367812#step:7:73)

**Review by yiheng-wang-nv — COMMENTED** [review]
For the `models/model.pt`, could it be replaced by providing a yml file which contains the download link? Here is an example: https://github.com/Project-MONAI/model-zoo/blob/dev/models/brats_mri_segmentation/large_files.yml

File: models/breast_density_classification/docs/README.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+# Sample Data
+In the folder `sample_data` few example input images are stored for each category of images. These images are stored in jpeg format for sharing purpose.
+
+# Input and Output Formats
+The input image should have the size [299, 299, 3]. For a dicom image which are single channel. The channel can be repeated 3 times.
+The output is an array with probabilities for each of the four class.
+
+# Sample Data
+In the folder `sample_data` few example input images are stored for each category of images. These images are stored in jpeg format for sharing purpose.
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/docs/README.md]: this section is duplicate, may need to be removed.

Comment (vikashg) [reply models/breast_density_classification/docs/README.md]: Thanks @yiheng-wang-nv  I will do that tomorrow morning and incorporate the changes you suggested. 

Vikash

File: models/breast_density_classification/scripts/create_dataset.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            _out = {"image": os.path.join(data_dir, _file), "label": _label}
+            output_list.append(_out)
+
+    data_dict = {"Test": output_list}
+
+    fid = open(output_file, "w")
+    json.dump(data_dict, fid, indent=1)
+
+
+if __name__ == "__main__":
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/scripts/create_dataset.py:34]: HI @vikashg , FYI, I added an arg parser for this file, thus users can type in the input and out path.

File: models/breast_density_classification/scripts/createList.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    test_data = file_dict["Test"][0:20]
+
+    data_dict = {}
+    data_dict["Train"] = train_data
+    data_dict["Validation"] = validation_data
+    data_dict["Test"] = test_data
+
+
+def main():
+    json_filelist = "/raid/Data/MayoClinicData/BreastDensity/patient_img_label_FFDM.json"
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/scripts/createList.py]: Hi @vikashg , for this file, can I ask if the original dataset is publicly accessible? If it is only for internal use, maybe this file is not needed.

File: models/breast_density_classification/scripts/script.py (hunk 1 above)
Comment (yiheng-wang-nv) [reply models/breast_density_classification/scripts/script.py:1]: I pushed some commits and the format issue has been resolved.

**Comment by yiheng-wang-nv** [comment]
HI @vikashg , I added some comments above, and for the format issue I already solved via pushing some commits.
Could you please help to check other comments, thanks!

File: models/breast_density_classification/scripts/createList.py (hunk 4 above)
Comment (vikashg) [reply models/breast_density_classification/scripts/createList.py]: Sorry that was the trace left from my testing and I am incorporating all your feed back now and I will soon update. 
Thanks

File: models/breast_density_classification/scripts/create_dataset.py (hunk 3 above)
Comment (vikashg) [reply models/breast_density_classification/scripts/create_dataset.py:34]: Thanks, I will update the documentation accordingly

**Comment by vikashg** [comment]
Hi @yiheng-wang-nv, I updated the changes you suggested. I need to add the Yaml file. However, I run into few warnings which are probably related to MONAI 1.0 meta tensor. I do not have much experience with meta tensor at the moment. So, if you can please look it will be great. Here are the warnings that are generated 
```
2022-10-10 11:46:13,113 - INFO - > logging_file: 'configs/logging.conf'
2022-10-10 11:46:13,113 - INFO - ---


2022-10-10 11:46:13,113 - INFO - set logging properties based on config: configs/logging.conf.
Loading dataset: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 16/16 [00:02<00:00,  6.97it/s]
/home/gupta/miniconda3/envs/monai/lib/python3.8/site-packages/torchvision/models/_utils.py:208: UserWarning: The parameter 'pretrained' is deprecated since 0.13 and will be removed in 0.15, please use 'weights' instead.
  warnings.warn(
/home/gupta/miniconda3/envs/monai/lib/python3.8/site-packages/torchvision/models/_utils.py:223: UserWarning: Arguments other than a weight enum or `None` for 'weights' are deprecated since 0.13 and will be removed in 0.15. The current behavior is equivalent to passing `weights=None`.
  warnings.warn(msg)
/home/gupta/miniconda3/envs/monai/lib/python3.8/site-packages/torchvision/models/inception.py:43: FutureWarning: The default weight initialization of inception_v3 will be changed in future releases of torchvision. If you wish to keep the old behavior (which leads to long initialization times due to scipy/scipy#11299), please set init_weights=True.
  warnings.warn(
2022-10-10 11:46:16,718 - ignite.engine.engine.SupervisedEvaluator - INFO - Engine run resuming from iteration 0, epoch 0 until 1 epochs
2022-10-10 11:46:16,811 - ignite.engine.engine.SupervisedEvaluator - INFO - Restored all variables from ./models/model.pt
/home/gupta/miniconda3/envs/monai/lib/python3.8/site-packages/monai/handlers/stats_handler.py:237: UserWarning: ignoring non-scalar output in StatsHandler, make sure `output_transform(engine.state.output)` returns a scalar or dictionary of key and scalar pairs to avoid this warning. image:<class 'monai.data.meta_tensor.MetaTensor'>
  warnings.warn(
/home/gupta/miniconda3/envs/monai/lib/python3.8/site-packages/monai/handlers/stats_handler.py:237: UserWarning: ignoring non-scalar output in StatsHandler, make sure `output_transform(engine.state.output)` returns a scalar or dictionary of key and scalar pairs to avoid this warning. label:<class 'NoneType'>
  warnings.warn(
/home/gupta/miniconda3/envs/monai/lib/python3.8/site-packages/monai/handlers/stats_handler.py:237: UserWarning: ignoring non-scalar output in StatsHandler, make sure `output_transform(engine.state.output)` returns a scalar or dictionary of key and scalar pairs to avoid this warning. pred:<class 'monai.data.meta_tensor.MetaTensor'>
  warnings.warn(
2022-10-10 11:46:19,285 - ignite.engine.engine.SupervisedEvaluator - INFO - Key metric: None best value: -1 at epoch: -1
2022-10-10 11:46:19,285 - ignite.engine.engine.SupervisedEvaluator - INFO - Epoch[1] Complete. Time taken: 00:00:02
2022-10-10 11:46:19,285 - ignite.engine.engine.SupervisedEvaluator - INFO - Engine run complete. Time taken: 00:00:03
[null, null]
```

I believe, it is in relation to how the present  `configs/inference.json` is handling the post-processing like `Activationsd`. I am looking into it. 
Thanks

File: models/breast_density_classification/configs/inference.json [hunk 5]
```diff
@@ -91,7 +92,14 @@
         },
         {
             "_target_": "StatsHandler",
-            "iteration_log": true
+            "iteration_log": false,
+            "output_transform": "$lambda x: None"
+        },
+        {
+            "_target_": "ClassificationSaver",
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/configs/inference.json:96]: Hi @vikashg , I pushed a new commit (https://github.com/Project-MONAI/model-zoo/pull/150/commits/74c44e34e4f14efa4c4d765d570024ad670cda97), and the `ClassificationSaver` is used to save the predictions into a csv file, you can check `monai.handlers.CheckpointSaver` for more details.

File: models/breast_density_classification/docs/README.md [hunk 6]
```diff
@@ -20,14 +20,14 @@ The output is an array with probabilities for each of the four class.
 # Commands Example
 Create a json file with names of all the input files. Execute the following command
 ```
-python scripts/create_dataset.py -base_dir <absolute path of data dir> --output_file ../configs/sample_image_data.json
+python scripts/create_dataset.py -base_dir <path to the bundle root dir>/sample_data -output_file configs/sample_image_data.json
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/docs/README.md:22]: The changes of the readme file is to unify the format with other bundles

File: models/breast_density_classification/configs/metadata.json [hunk 7]
```diff
@@ -1,10 +1,10 @@
 {
     "version": "0.1.0",
-    "monai_version": "1.0.0rc1",
+    "monai_version": "1.0.0",
     "pytorch_version": "1.10.0",
     "numpy_version": "1.21.2",
     "optional_packages_version": {
-        "pydicom": "3.2.1"
+        "torchvision": "0.11.0"
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/configs/metadata.json]: It seems pydicom is not used, thus I removed it. Fix me if it may be useful. I also added torchvision and changed the monai version into a formal release version.

Comment (vikashg) [reply models/breast_density_classification/configs/metadata.json]: I kept the pydicom in order to ensure that if someone uses a dicom image.

File: models/breast_density_classification/configs/inference.json (hunk 5 above)
Comment (vikashg) [reply models/breast_density_classification/configs/inference.json:96]: Awesome Thanks. Now I can see the saved csv file.

**Comment by vikashg** [comment]
Thanks @yiheng-wang-nv I tested it works now. I will work on creating a training script. I guess you can merge now right ?

**Comment by vikashg** [comment]
Oh I remember I have to upload the model on google drive and share a link. I will do that.

**Comment by vikashg** [comment]
@yiheng-wang-nv Is there anything else I need to do before this is merged. I was looking into creating the large_files.yaml file. How can I generate the hash value ?

**Comment by yiheng-wang-nv** [comment]
> @yiheng-wang-nv Is there anything else I need to do before this is merged. I was looking into creating the large_files.yaml file. How can I generate the hash value ?

Hi @vikashg , do you use linux system? If so, you can run `md5sum model.pt` to get the hash value for `model.pt`

**Comment by vikashg** [comment]
> > @yiheng-wang-nv Is there anything else I need to do before this is merged. I was looking into creating the large_files.yaml file. How can I generate the hash value ?
> 
> Hi @vikashg , do you use linux system? If so, you can run `md5sum model.pt` to get the hash value for `model.pt`

I have added the large_files.yml file.

**Comment by yiheng-wang-nv** [comment]
> > > @yiheng-wang-nv Is there anything else I need to do before this is merged. I was looking into creating the large_files.yaml file. How can I generate the hash value ?
> > 
> > 
> > Hi @vikashg , do you use linux system? If so, you can run `md5sum model.pt` to get the hash value for `model.pt`
> 
> I have added the large_files.yml file.

Thanks @vikashg , I pushed a new commit (see: https://github.com/Project-MONAI/model-zoo/pull/150/commits/95a027f87c596306345595f51b14ad75f4531031) in order to ensure necessary parts are included in `metadata.json` and `inference.json`. Could you please double check it?

File: models/breast_density_classification/docs/README.md [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+```
+
+# Execute Inference
+The inference can be executed as follows
+```
+python -m monai.bundle run evaluating --meta_file configs/metadata.json --config_file configs/inference.json configs/logging.conf
+```
+
+# Execute training
+It is a work in progress and will be shared soon.
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/docs/README.md]: HI @vikashg , should this sentence be: "It is a work in progress and will be shared in the next version soon" ?

File: models/breast_density_classification/configs/metadata.json [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    },
+    "task": "Breast Density Classification",
+    "description": "A pre-trained model for classifying breast images (mammograms)  ",
+    "authors": "Center for Augmented Intelligence in Imaging, Mayo Clinic Florida",
+    "copyright": "Copyright (c) Mayo Clinic",
+    "data_source": "Mayo Clinic ",
+    "data_type": "Jpeg",
+    "image_classes": "three channel data, intensity scaled to [0, 1]. A single grayscale is copied to 3 channels",
+    "label_classes": "four classes marked as [0, 1, 0, 0]",
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/configs/metadata.json]: here may need to describe more clearly, like `[1, 0, 0, 0]` for A, `[0, 1, 0, 0]` for B ...

File: models/breast_density_classification/docs/README.md [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+A pre-trained model for breast-density classification.
+
+# Model Overview
+This model is trained using transfer learning on InceptionV3. The model weights were fine tuned using the Mayo Clinic Data. The details of training and data is outlined in https://arxiv.org/abs/2202.08238.
+
+The bundle does not support torchscript.
+
+# Input and Output Formats
+The input image should have the size [299, 299, 3]. For a dicom image which are single channel. The channel can be repeated 3 times.
+The output is an array with probabilities for each of the four class.
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/docs/README.md]: HI @vikashg , could you please double confirm here? If the output is like this, `AsDiscreted` can be removed from `postprocessing` and we only need to keep the sigmoid.

File: models/breast_density_classification/large_files.yml [hunk 11]
```diff
@@ -0,0 +1,5 @@
+large_files:
+  - path: "models/model.pt"
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/large_files.yml:2]: Could you please help to double check the weights? When I run the inference script, all outputs are [1, 0, 0, 0]

File: models/breast_density_classification/configs/metadata.json [hunk 12]
```diff
@@ -0,0 +1,75 @@
+{
+    "schema": "https://github.com/Project-MONAI/MONAI-extra-test-data/releases/download/0.8.1/meta_schema_20220324.json",
+    "changelog": {
+        "0.1.0": "complete the model package"
+    },
+    "version": "0.1.0",
+    "monai_version": "1.0.0",
+    "pytorch_version": "1.10.0",
```
Comment (yiheng-wang-nv) [inline models/breast_density_classification/configs/metadata.json]: Hi @vikashg , ci test failed due to lib version conflicts. It seems in order to install an old torch vision as well as torchvision, cuda version is required (see: https://pytorch.org/get-started/previous-versions/#linux-and-windows-5) Could you please help to double check the pytorch version, torchvision version and cuda version? Thanks!

File: models/breast_density_classification/docs/README.md (hunk 10 above)
Comment (vikashg) [reply models/breast_density_classification/docs/README.md]: ok I can do that

File: models/breast_density_classification/configs/metadata.json (hunk 9 above)
Comment (vikashg) [reply models/breast_density_classification/configs/metadata.json]: ok I will do that.

File: models/breast_density_classification/large_files.yml (hunk 11 above)
Comment (vikashg) [reply models/breast_density_classification/large_files.yml:2]: Yes this is a part that is of concern to me as well and my likely conclusion is that I need to retrain the model as the previous model that I had was trained with monai-0.9 cannot be loaded with the new monai-1.0 which supports the TorchVisionFC module. I am working on retraining the model. So, I pushed a model which was trained for only a few epochs to ensure that the pipeline is working and the idea was to replace the model with the newly trained one once it is complete.

File: models/breast_density_classification/configs/metadata.json (hunk 12 above)
Comment (vikashg) [reply models/breast_density_classification/configs/metadata.json]: I think it was a residual from the earlier versions. Changed it.

**Comment by vikashg** [comment]
> /build

?

**Comment by yiheng-wang-nv** [comment]
> > /build
> 
> ?

Hi @vikashg , this command is used to trigger the blossom ci test. This PR now looks fine, and after uploading the trained weights, It's able to merge, thanks!

**Review by wyli — APPROVED** [review]
Thanks, it looks good to me.

**Comment by vikashg** [comment]
Hi @yiheng-wang-nv 
I uploaded a new trained model and the results good to me. Can you have a look and merge it ?

image

**Review by vikashg — COMMENTED** [review]
Looks good to me.

**Comment by yiheng-wang-nv** [comment]
Hi @vikashg , thanks for the updates. The new weights works fine. A minor issue is that could you please modify the weights name in google drive from `model.pth` into `model.pt`? We require to unify the naming format for bundles (https://github.com/Project-MONAI/model-zoo/blob/dev/CONTRIBUTING.md#prepare-a-config-file-for-large-files). Since it is not related to the bundle side, and the renaming will not change the hash value, I will merge this PR first. Thanks again for your efforts!
